File: application/api/forms/model-grid-column.js
/* global Java */

define(['boxing', 'common-utils/color', 'common-utils/cursor', 'common-utils/font', './cell-render-event'], function(B, Color, Cursor, Font, RenderEvent) {
    var className = "com.eas.client.forms.components.model.grid.header.ModelGridColumn";
    var javaClass = Java.type(className);
    /**
     *
     * @constructor ModelGridColumn ModelGridColumn
     */
    function ModelGridColumn() {
        var maxArgs = 0;
        var delegate = arguments.length > maxArgs ?
              arguments[maxArgs] 
            : new javaClass();

        Object.defineProperty(this, "unwrap", {
            configurable: true,
            value: function() {
                return delegate;
            }
        });
        if(ModelGridColumn.superclass)
            ModelGridColumn.superclass.constructor.apply(this, arguments);
        delegate.setPublished(this);
        this.preferredWidth = 0;
        Object.defineProperty(this, "preferredWidth", {
            get: function() {
                var value = delegate.preferredWidth;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.preferredWidth = B.boxAsJava(aValue);
            }
        });

        this.visible = true;
        Object.defineProperty(this, "visible", {
            get: function() {
                var value = delegate.visible;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.visible = B.boxAsJava(aValue);
            }
        });

        this.resizable = true;
        Object.defineProperty(this, "resizable", {
            get: function() {
                var value = delegate.resizable;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.resizable = B.boxAsJava(aValue);
            }
        });

        this.minWidth = 0;
        Object.defineProperty(this, "minWidth", {
            get: function() {
                var value = delegate.minWidth;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.minWidth = B.boxAsJava(aValue);
            }
        });

        this.foreground = new Object();
        Object.defineProperty(this, "foreground", {
            get: function() {
                var value = delegate.foreground;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.foreground = B.boxAsJava(aValue);
            }
        });

        this.sortable = true;
        Object.defineProperty(this, "sortable", {
            get: function() {
                var value = delegate.sortable;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.sortable = B.boxAsJava(aValue);
            }
        });

        this.title = '';
        Object.defineProperty(this, "title", {
            get: function() {
                var value = delegate.title;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.title = B.boxAsJava(aValue);
            }
        });

        /**
         * Returns script handler, used for select a value of the cell.
         */
        this.onSelect = new Object();
        Object.defineProperty(this, "onSelect", {
            get: function() {
                var value = delegate.onSelect;
                return value;
            },
            set: function(aValue) {
                delegate.onSelect = aValue;
            }
        });

        this.movable = true;
        Object.defineProperty(this, "movable", {
            get: function() {
                var value = delegate.movable;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.movable = B.boxAsJava(aValue);
            }
        });

        this.field = '';
        Object.defineProperty(this, "field", {
            get: function() {
                var value = delegate.field;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.field = B.boxAsJava(aValue);
            }
        });

        this.readonly = true;
        Object.defineProperty(this, "readonly", {
            get: function() {
                var value = delegate.readonly;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.readonly = B.boxAsJava(aValue);
            }
        });

        /**
         * Returns script handler, used for calculate cell's data, display value and style attributes.
         */
        this.onRender = new Object();
        Object.defineProperty(this, "onRender", {
            get: function() {
                var value = delegate.onRender;
                return value;
            },
            set: function(aValue) {
                delegate.onRender = aValue;
            }
        });

        this.background = new Object();
        Object.defineProperty(this, "background", {
            get: function() {
                var value = delegate.background;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.background = B.boxAsJava(aValue);
            }
        });

        this.width = 0;
        Object.defineProperty(this, "width", {
            get: function() {
                var value = delegate.width;
                return B.boxAsJs(value);
            }
        });

        this.sortField = '';
        Object.defineProperty(this, "sortField", {
            get: function() {
                var value = delegate.sortField;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.sortField = B.boxAsJava(aValue);
            }
        });

        this.font = new Object();
        Object.defineProperty(this, "font", {
            get: function() {
                var value = delegate.font;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.font = B.boxAsJava(aValue);
            }
        });

        this.maxWidth = 0;
        Object.defineProperty(this, "maxWidth", {
            get: function() {
                var value = delegate.maxWidth;
                return B.boxAsJs(value);
            },
            set: function(aValue) {
                delegate.maxWidth = B.boxAsJava(aValue);
            }
        });

    }
    /**
     * Column sort, works only in HTML5
     * @method sort
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.sort = function() {
        var delegate = this.unwrap();
        var value = delegate.sort();
        return B.boxAsJs(value);
    };

    /**
     * Clears sort column, works only in HTML5
     * @method unsort
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.unsort = function() {
        var delegate = this.unwrap();
        var value = delegate.unsort();
        return B.boxAsJs(value);
    };

    /**
     *
     * @method addColumnNode
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.addColumnNode = function(node) {
        var delegate = this.unwrap();
        var value = delegate.addColumnNode(B.boxAsJava(node));
        return B.boxAsJs(value);
    };

    /**
     *
     * @method columnNodes
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.columnNodes = function() {
        var delegate = this.unwrap();
        var value = delegate.columnNodes();
        return B.boxAsJs(value);
    };

    /**
     *
     * @method removeColumnNode
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.removeColumnNode = function(node) {
        var delegate = this.unwrap();
        var value = delegate.removeColumnNode(B.boxAsJava(node));
        return B.boxAsJs(value);
    };

    /**
     *
     * @method insertColumnNode
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.insertColumnNode = function(position, node) {
        var delegate = this.unwrap();
        var value = delegate.insertColumnNode(B.boxAsJava(position), B.boxAsJava(node));
        return B.boxAsJs(value);
    };

    /**
     * Descending column sort, works only in HTML5
     * @method sortDesc
     * @memberOf ModelGridColumn
     */
    ModelGridColumn.prototype.sortDesc = function() {
        var delegate = this.unwrap();
        var value = delegate.sortDesc();
        return B.boxAsJs(value);
    };


    var ScriptsClass = Java.type("com.eas.script.Scripts");
    var space = ScriptsClass.getSpace();
    space.putPublisher(className, function(aDelegate) {
        return new ModelGridColumn(aDelegate);
    });
    return ModelGridColumn;
});

File: platypus-js-forms/src/main/java/com/eas/client/forms/components/model/grid/header/ModelGridColumn.java
package com.eas.client.forms.components.model.grid.header;

import com.bearsoft.gui.grid.header.GridColumnsNode;
import com.eas.client.forms.components.model.ModelWidget;
import com.eas.client.forms.components.model.grid.columns.ModelColumn;
import com.eas.design.Designable;
import com.eas.design.Undesignable;
import com.eas.script.AlreadyPublishedException;
import com.eas.script.HasPublished;
import com.eas.script.NoPublisherException;
import com.eas.script.ScriptFunction;
import com.eas.script.Scripts;
import java.awt.Color;
import java.awt.Font;
import jdk.nashorn.api.scripting.JSObject;

/**
 *
 * @author Марат
 */
public class ModelGridColumn extends GridColumnsNode implements HasPublished {

    protected JSObject published;

    protected String field;
    protected String sortField;

    @ScriptFunction
    public ModelGridColumn() {
        super();
        setTableColumn(new ModelColumn());
    }

    @Override
    public GridColumnsNode lightCopy() {
        GridColumnsNode copied = new ModelGridColumn();
        copied.lightAssign(this);
        return copied;
    }

    @Override
    public GridColumnsNode copy() throws Exception {
        GridColumnsNode copied = new ModelGridColumn();
        copied.assign(this);
        return copied;
    }

    @ScriptFunction(jsDoc = ""
            + "/**\n"
            + " * Returns script handler, used for calculate cell's data, display value and style attributes.\n"
            + " */")
    public JSObject getOnRender() {
        return ((ModelColumn) getTableColumn()).getOnRender();
    }

    @ScriptFunction
    public void setOnRender(JSObject aValue) {
        ((ModelColumn) getTableColumn()).setOnRender(aValue);
    }

    @ScriptFunction(jsDoc = ""
            + "/**\n"
            + " * Returns script handler, used for select a value of the cell.\n"
            + " */")
    public JSObject getOnSelect() {
        return ((ModelColumn) getTableColumn()).getOnSelect();
    }

    @ScriptFunction
    public void setOnSelect(JSObject aValue) throws Exception {
        ((ModelColumn) getTableColumn()).setOnSelect(aValue);
    }

    @Undesignable
    @ScriptFunction
    public ModelWidget getEditor() {
        return ((ModelColumn) getTableColumn()).getEditor();
    }

    @ScriptFunction
    public void setEditor(ModelWidget aEditor) {
        ((ModelColumn) getTableColumn()).setEditor(aEditor);
    }

    @Undesignable
    @ScriptFunction
    public ModelWidget getView() {
        return ((ModelColumn) getTableColumn()).getView();
    }

    @ScriptFunction
    public void setView(ModelWidget aEditor) {
        ((ModelColumn) getTableColumn()).setView(aEditor);
    }

    @ScriptFunction(params = {"node"})
    @Override
    public void removeColumnNode(GridColumnsNode aNode) {
        super.removeColumnNode(aNode);
    }

    @ScriptFunction(params = {"node"})
    @Override
    public void addColumnNode(GridColumnsNode aNode) {
        super.addColumnNode(aNode);
    }

    @ScriptFunction(params = {"position", "node"})
    @Override
    public void insertColumnNode(int atIndex, GridColumnsNode aNode) {
        super.insertColumnNode(atIndex, aNode);
    }

    @ScriptFunction
    public GridColumnsNode[] columnNodes() {
        return children.toArray(new GridColumnsNode[]{});
    }

    @ScriptFunction
    @Designable(category = "model")
    public String getField() {
        return field;
    }

    @ScriptFunction
    public void setField(String aValue) {
        if (field == null ? aValue != null : !field.equals(aValue)) {
            field = aValue;
            ((ModelColumn) tableColumn).setField(field);
        }
    }

    @ScriptFunction
    @Designable(category = "model")
    public String getSortField() {
        return sortField;
    }

    @ScriptFunction
    public void setSortField(String aValue) {
        if (sortField == null ? aValue != null : !sortField.equals(aValue)) {
            sortField = aValue;
            ((ModelColumn) tableColumn).setSortField(sortField);
        }
    }

    @ScriptFunction
    @Override
    public int getWidth() {
        return super.getWidth();
    }

    @Override
    public void setWidth(int aValue) {
        super.setWidth(aValue);
    }

    @ScriptFunction
    @Override
    public int getPreferredWidth() {
        return super.getPreferredWidth();
    }

    @ScriptFunction
    @Override
    public void setPreferredWidth(int aValue) {
        super.setPreferredWidth(aValue);
    }

    @ScriptFunction
    @Override
    public int getMinWidth() {
        return super.getMinWidth();
    }

    @ScriptFunction
    @Override
    public void setMinWidth(int aValue) {
        super.setMinWidth(aValue);
    }

    @ScriptFunction
    @Override
    public int getMaxWidth() {
        return super.getMaxWidth();
    }

    @ScriptFunction
    @Override
    public void setMaxWidth(int aValue) {
        super.setMaxWidth(aValue);
    }

    @ScriptFunction
    @Override
    public boolean isReadonly() {
        return super.isReadonly();
    }

    @ScriptFunction
    @Override
    public void setReadonly(boolean aValue) {
        super.setReadonly(aValue);
    }

    @ScriptFunction
    @Override
    public boolean isVisible() {
        return super.isVisible();
    }

    @ScriptFunction
    @Override
    public void setVisible(boolean aValue) {
        super.setVisible(aValue);
    }

    @ScriptFunction
    @Override
    public boolean isMovable() {
        return super.isMovable();
    }

    @ScriptFunction
    @Override
    public void setMovable(boolean aValue) {
        super.setMovable(aValue);
    }

    @ScriptFunction
    @Override
    public boolean isResizable() {
        return super.isResizable();
    }

    @ScriptFunction
    @Override
    public void setResizable(boolean aValue) {
        super.setResizable(aValue);
    }

    @ScriptFunction
    @Override
    public boolean isSortable() {
        return super.isSortable();
    }

    @ScriptFunction
    @Override
    public void setSortable(boolean aValue) {
        super.setSortable(aValue);
    }

    @ScriptFunction
    @Override
    public String getTitle() {
        return super.getTitle();
    }

    @ScriptFunction
    @Override
    public void setTitle(String aTitle) {
        super.setTitle(aTitle);
    }

    @ScriptFunction
    @Override
    public Color getBackground() {
        return super.getBackground();
    }

    @ScriptFunction
    @Override
    public void setBackground(Color aValue) {
        super.setBackground(aValue);
    }

    @ScriptFunction
    @Override
    public Color getForeground() {
        return super.getForeground();
    }

    @ScriptFunction
    @Override
    public void setForeground(Color aValue) {
        super.setForeground(aValue);
    }

    @ScriptFunction
    @Override
    public Font getFont() {
        return super.getFont();
    }

    @ScriptFunction
    @Override
    public void setFont(Font aValue) {
        super.setFont(aValue);
    }

    @Override
    public JSObject getPublished() {
        if (published == null) {
            JSObject publisher = Scripts.getSpace().getPublisher(this.getClass().getName());
            if (publisher == null || !publisher.isFunction()) {
                throw new NoPublisherException();
            }
            published = (JSObject) publisher.call(null, new Object[]{this});
        }
        return published;
    }

    @Override
    public void setPublished(JSObject jsColumn) {
        if (published != null) {
            throw new AlreadyPublishedException();
        }
        published = jsColumn;
        ((ModelColumn) getTableColumn()).setEventsSource(jsColumn);
        /*
         if(view != null)
         view.injectPublished(published);
         if(editor != null)
         editor.injectPublished(published);
         */
    }

    private static final String SORT = ""
            + "/**\n"
            + " * Column sort, works only in HTML5\n"
            + " */";

    @ScriptFunction(jsDoc = SORT)
    public void sort() {

    }
    private static final String SORT_DESC = ""
            + "/**\n"
            + " * Descending column sort, works only in HTML5\n"
            + " */";

    @ScriptFunction(jsDoc = SORT_DESC)
    public void sortDesc() {

    }
    private static final String UNSORT = ""
            + "/**\n"
            + " * Clears sort column, works only in HTML5\n"
            + " */";

    @ScriptFunction(jsDoc = UNSORT)
    public void unsort() {

    }
}
